### PR TITLE
dev-qt/*:6 enable tests and other misc improvements

### PR DESCRIPTION
**Pull request by ionenwks**
Tests should pass on amd64 in theory (x86 too minus webengine, not that I tried much w/ x86), but I do imagine there will be edge cases to cover.

See commit messages for details.

Note qtwebengine builds and passes tests on llvm-musl too.

**Edit:**
Mostly ready at this point but may still do a few small things / tests before merge.

Things not (really) worked on yet (may not get to everything before merge):
- ~figuring out a sane solution to the intrinsics mess~
- ~qtwebengine (known failing with qtwebchannel[-qml] but that's because did not make quick optional yet)~
- ~retest recent changes with minimal USE/deps~
- ~checking open bugs~
- ~rechecking my own commits for obvious mistakes~
- ~checking if anything needs [gles2-only=], need to recheck if still missing any vulkan=,opengl= too~
- ~trying live ebuilds~
- ~try musl, libc++~

Things that considered but not going to get to in this PR:
- examples (do kind of want this one eventually given nice for testing)
- docs
- qt-creator (albeit, along with the two above, must say my focus is more to ensure working qt6 stuff for other packages than aid qt developers -- not to say it doesn't deserve attention)
- enable pdf module on qtwebengine (probably easy, but don't want more rebuilds to figure that one out just yet)
- trying and fixing cross
- licenses (eclass' licenses is just not enough for this, there need to be per packages additions)
- split qtwaylandscanner? not entirely sure even want to do this one anymore, there are plenty of other small tools needed for cross (e.g. qsb from qtshader) and overall will be a pain to split, avoid conflicts (all needs sed/patches), run tests, and then keep maintaining that -- just so rare cross users have a bit less to emerge on CBUILD. qtwebengine's gn could make more sense though and "think" its cmake switches can do a clean split.

~Decent odds will end up adding 6.5.3 bump before merge this given it's steadily approaching (Sep 5 if on schedule).~ Or looks(?) like more like Sep 19+, so maybe merge earlier.

Merging will cause some degree of disruption (revbumps, matching opengl/vulkan IUSE for those that aren't using global or desktop profiles, potentially affect some revdeps due to more things being optional)

### CODE REVIEW

File: dev-qt/qtbase/qtbase-6.5.2-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	"${FILESDIR}"/${P}-CVE-2023-38197.patch
+	"${FILESDIR}"/${P}-tests-gcc13.patch
+)
+
+src_prepare() {
+	qt6-build_src_prepare
+
+	if use test; then
+		# test itself has -Werror=strict-aliasing issues, drop for simplicity
+		sed -e '/add_subdirectory(qsharedpointer)/d' \
```
Comment (thesamesam) [inline dev-qt/qtbase/qtbase-6.5.2-r1.ebuild:135]: I don't ever use it but could use `cmake_comment_add_subdirectory`?

File: dev-qt/qtmultimedia/qtmultimedia-6.5.2.ebuild [hunk 2]
```diff
@@ -34,30 +33,45 @@ RDEPEND="
...
+	tst_qmediacapturesession
+	tst_qmediaplayerbackend
+	tst_qsoundeffect
+	# may try to use hardware acceleration for encoding
+	tst_qscreencapture_integration
+)
+
 src_configure() {
+	# normally passed by build system, but needed for 32-on-64 chroots
+	use x86 && append-cppflags -DPFFFT_SIMD_DISABLE
```
Comment (thesamesam) [inline dev-qt/qtmultimedia/qtmultimedia-6.5.2.ebuild:67]: ha, I googled it because I was wondering wtf it stood for ("FFT", ok, but the rest?) and came across https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/pffft/patches/05-fix-arch-detection.diff.

File: dev-qt/qttools/qttools-6.9999.ebuild [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	assistant? ( =dev-qt/qtbase-${PV}*:6[sql,widgets] )
+	designer? ( =dev-qt/qtbase-${PV}*:6[widgets] )
+	distancefieldgenerator? (
+		=dev-qt/qtbase-${PV}*:6[widgets]
+		=dev-qt/qtdeclarative-${PV}*:6
+	)
+	pixeltool? ( =dev-qt/qtbase-${PV}*:6[widgets] )
+	qdbus? ( =dev-qt/qtbase-${PV}*:6[widgets] )
+	qdiag? ( =dev-qt/qtbase-${PV}*:6[opengl,widgets] )
+	qdoc? ( sys-devel/clang:= )
```
Comment (thesamesam) [inline dev-qt/qttools/qttools-6.9999.ebuild:35]: Do we consider libclang usage without llvm.eclass to be smelly?

File: eclass/qt6-build.eclass [hunk 4]
```diff
@@ -19,111 +19,143 @@ if [[ -z ${_QT6_BUILD_ECLASS} ]]; then
...
+	else
+		local mycmakeargs=()
+	fi
+
+	mycmakeargs+=(
+		# note that if qtbase was built with tests, this is default ON
+		-DQT_BUILD_TESTS=$(in_iuse test && usev test ON || echo OFF)
+	)
+
+	[[ ${PN} != qttranslations ]] &&
```
Comment (thesamesam) [inline eclass/qt6-build.eclass:114]: what's the story with the conditional?

File: eclass/qt6-build.eclass [hunk 5]
```diff
@@ -19,111 +19,143 @@ if [[ -z ${_QT6_BUILD_ECLASS} ]]; then
...
+		mycmakeargs+=( -DQT_USE_DEFAULT_CMAKE_OPTIMIZATION_FLAGS=ON )
+
+	# LTO cause test failures in several components (e.g. qtcharts,
+	# multimedia, scxml, wayland, webchannel, ...).
+	#
+	# Exact extend/causes unknown, but for some related-sounding bugs:
```
Comment (thesamesam) [inline eclass/qt6-build.eclass:120]: Maybe:
```suggestion
	# Exact extent/causes unknown, but for some related-sounding bugs:
```

File: dev-qt/qtwebengine/qtwebengine-6.9999.ebuild [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	local chromium
+	mapfile -t chromium < CHROMIUM_VERSION || die
+	[[ ${chromium[1]} =~ ^Based.*:[^0-9]+([0-9.]+$) ]] &&
+		QT6_CHROMIUM_VER=${BASH_REMATCH[1]} || die
+	[[ ${chromium[2]} =~ ^Patched.+:[^0-9]+([0-9.]+$) ]] &&
+		QT6_CHROMIUM_PATCHES_VER=${BASH_REMATCH[1]} || die
+}
+
+src_configure() {
+	export NINJA_PATH=${BROOT}/usr/bin/ninja
```
Comment (thesamesam) [inline dev-qt/qtwebengine/qtwebengine-6.9999.ebuild:173]: Wonder if should allow ${NINJA} override w/ type -P or something but I'm not sure if we really care about testing it with samu.

File: dev-qt/qtwebengine/metadata.xml [hunk 7]
```diff
@@ -20,6 +20,7 @@
 	<upstream>
 		<bugs-to>https://bugreports.qt.io/</bugs-to>
 		<doc>https://doc.qt.io/</doc>
+		<remote-id type="github">qt/qtwebengine</remote-id>
```
Comment (thesamesam) [inline dev-qt/qtwebengine/metadata.xml:23]: chuck in https://github.com/qt/qtwebengine-chromium

File: dev-qt/qttools/qttools-6.9999.ebuild (hunk 3 above)
Comment (ionenwks) [reply dev-qt/qttools/qttools-6.9999.ebuild:35]: That's already in my TODO (I didn't get to qttools yet), not marked as draft for nothing

File: dev-qt/qtbase/qtbase-6.5.2-r1.ebuild (hunk 1 above)
Comment (ionenwks) [reply dev-qt/qtbase/qtbase-6.5.2-r1.ebuild:135]: Considered it already but that function only edit the CMakeLists.txt in current directory, so I'd have to push/pop which doesn't feel much better.

File: eclass/qt6-build.eclass (hunk 4 above)
Comment (ionenwks) [reply eclass/qt6-build.eclass:114]: It's noted in the commit message, but the tl;dr is that it doesn't compile anything so it gives a unused cmake option warning. I could add a comment fwiw

File: dev-qt/qtwebengine/qtwebengine-6.9999.ebuild (hunk 6 above)
Comment (ionenwks) [reply dev-qt/qtwebengine/qtwebengine-6.9999.ebuild:173]: In my TODO I have a line about checking if I can just remove that altogether. I think this may be a remnant from the Qt5 ebuild (pre-cmake). Like qttools I didn't really look at qtwebengine yet.

File: dev-qt/qttools/qttools-6.9999.ebuild (hunk 3 above)
Comment (ionenwks) [reply dev-qt/qttools/qttools-6.9999.ebuild:35]: Also probably need to have a USE=clang/llvm or so, given clang can be used for other things it seems too and I don't think merging it with USE=qdoc will keep working out.

File: eclass/qt6-build.eclass [hunk 8]
```diff
@@ -133,48 +131,48 @@ qt6-build_src_install() {
...
 
 # @FUNCTION: qt6_symlink_binary_to_path
 # @USAGE: <target binary name> [suffix]
 # @DESCRIPTION:
-# Symlink a given binary from QT6_BINDIR to QT6_PREFIX/bin, with optional suffix
+# Symlink a given binary from QT6_BINDIR to QT6_PREFIX/bin, with
+# optional suffix.
 qt6_symlink_binary_to_path() {
-	[[ $# -ge 1 ]] || die "${FUNCNAME}() requires at least one argument"
+	[[ ${#} -ge 1 ]] || die "${FUNCNAME}() requires at least one argument"
```
Comment (Pesa) [inline eclass/qt6-build.eclass:179]: I don't really see the point of these stylistic changes, but whatever...

File: dev-qt/qtwebengine/qtwebengine-6.9999.ebuild (hunk 6 above)
Comment (ionenwks) [reply dev-qt/qtwebengine/qtwebengine-6.9999.ebuild:173]: Yeah, from a quick look NINJA_PATH is ignored:
```
find_package(Ninja 1.7.2 REQUIRED)
<snip>
    COMMAND Ninja::ninja -C ${GN_BINARY_DIR}/$<CONFIG> ${GN_EXECUTABLE}
```
Looking at FindNinja.cmake, if we wanted alternatives I could pass -DNinja_EXECUTABLE="something" though -- but it also does `ninja --version` and may not be happy with samu as-is.

As noted in one of the commit message, do wish could just use dev-util/gn but qtwebengine uses a modified gn (unsure how compatible it is) and also insists on taking it from qt6 directories if using system's.

File: eclass/qt6-build.eclass [hunk 9]
```diff
@@ -133,48 +131,48 @@ qt6-build_src_install() {
...
+	readonly QT6_DOCDIR=${QT6_PREFIX}/share/qt6-doc
+	readonly QT6_EXAMPLESDIR=${QT6_DATADIR}/examples
+	readonly QT6_HEADERDIR=${QT6_PREFIX}/include/qt6
+	readonly QT6_IMPORTDIR=${QT6_ARCHDATADIR}/imports
+	readonly QT6_LIBEXECDIR=${QT6_ARCHDATADIR}/libexec
+	readonly QT6_PLUGINDIR=${QT6_ARCHDATADIR}/plugins
+	readonly QT6_QMLDIR=${QT6_ARCHDATADIR}/qml
+	readonly QT6_SYSCONFDIR=${EPREFIX}/etc/xdg
+	readonly QT6_TESTSDIR=${QT6_DATADIR}/tests
+	readonly QT6_TRANSLATIONDIR=${QT6_DATADIR}/translations
```
Comment (Pesa) [inline eclass/qt6-build.eclass:209]: IIRC I wanted to use `declare -g -r` here but couldn't because we still had to support an older version of bash that didn't understand `-g`. That's no longer a concern now, though I dunno if using `declare` provides any advantages over `readonly`.

File: eclass/qt6-build.eclass [hunk 10]
```diff
@@ -19,13 +18,12 @@ if [[ -z ${_QT6_BUILD_ECLASS} ]]; then
 _QT6_BUILD_ECLASS=1
 
 [[ ${CATEGORY} != dev-qt ]] &&
-	die "${ECLASS} is only to be used for building Qt 6"
+	die "${ECLASS} is only to be used for building Qt6"
 
 # @ECLASS_VARIABLE: QT6_MODULE
-# @PRE_INHERIT
```
[the remark this answers is not in the document]
Comment (ionenwks) [reply eclass/qt6-build.eclass:25]: Thanks, that's unintended actually. I had formerly added `@PRE_INHERIT` to VIRTUALX_REQUIRED, but then ended up removing virtualx entirely and reverted that bit.. wrongly.

File: eclass/qt6-build.eclass [hunk 11]
```diff
@@ -37,53 +35,51 @@ _QT6_BUILD_ECLASS=1
 
 inherit cmake virtualx
 
-HOMEPAGE="https://www.qt.io/"
-LICENSE="|| ( GPL-2 GPL-3 LGPL-3 ) FDL-1.3"
-SLOT=6/$(ver_cut 1-2)
-
-QT6_MINOR_VERSION=$(ver_cut 2)
-readonly QT6_MINOR_VERSION
+readonly QT6_MINOR_VERSION=$(ver_cut 2)
```
[the remark this answers is not in the document]
Comment (ionenwks) [reply eclass/qt6-build.eclass]: Good to know, but yeah I don't think there's issues nowadays. It's also used that way in other ebuilds like cargo.eclass and java-ant-2.eclass

File: eclass/qt6-build.eclass [hunk 12]
```diff
@@ -37,53 +35,51 @@ _QT6_BUILD_ECLASS=1
...
 		;;
 	*_alpha*|*_beta*|*_rc*)
 		# development release
-		QT6_BUILD_TYPE="release"
-		MY_P=${QT6_MODULE}-everywhere-src-${PV/_/-}
-		SRC_URI="https://download.qt.io/development_releases/qt/${PV%.*}/${PV/_/-}/submodules/${MY_P}.tar.xz"
-		S=${WORKDIR}/${MY_P}
+		readonly QT6_BUILD_TYPE=release
+		S=${WORKDIR}/${QT6_MODULE}-everywhere-src-${PV/_/-}
+		SRC_URI="https://download.qt.io/development_releases/qt/${PV%.*}/${PV/_/-}/submodules/${S##*/}.tar.xz"
```
[the remark this answers is not in the document]
Comment (ionenwks) [reply eclass/qt6-build.eclass:47]: Global scope pollution, MY is especially off in eclass given that's typically reserved for ebuilds. I could use something like QT6_P I guess though.

File: eclass/qt6-build.eclass (hunk 8 above)
Comment (ionenwks) [reply eclass/qt6-build.eclass:179]: For consistency, the eclass already had a mix of both, e.g.
`dosym -r "${QT6_BINDIR}"/${1} /usr/bin/${1}${2}`
I don't mind either but I prefer it's all the same within one eclass/ebuild. Typically put ${} unconditionally, so picked that one.

File: eclass/qt6-build.eclass (hunk 9 above)
Comment (ionenwks) [reply eclass/qt6-build.eclass:209]: Nah, beside that I find readonly kind of ugly :) But declare would be longer, so...

File: eclass/qt6-build.eclass (hunk 12 above)
[the remark this answers is not in the document]
Comment (ionenwks) [reply eclass/qt6-build.eclass:47]: And yeah, think I'll unset it too given pkgcheck warns about MY_P not being documented :) (was planning to document the other as-is, but hadn't got to that yet)

[the remark this answers is not in the document]
Comment (ionenwks) [reply eclass/qt6-build.eclass:47]: Alternatively, there's also funny tricks:
```
        : "${QT6_MODULE}-everywhere-src-${PV}"
        SRC_URI="https://download.qt.io/official_releases/qt/${PV%.*}/${PV}/submodules/${_}.tar.xz" \
            S=${WORKDIR}/${_}
```

File: dev-qt/qtwebengine/qtwebengine-6.9999.ebuild (hunk 6 above)
Comment (ionenwks) [reply dev-qt/qtwebengine/qtwebengine-6.9999.ebuild:173]: After finding it out that NINJAFLAGS is ignored too (it uses it for a ninja command on windows, but not to build gn), I thought might as well try samu for :rocket:`SAMUFLAGS`:rocket: but what I feared did happen:
```
  Could NOT find Ninja: (Required is at least version "1.7.2") (found samu)
```

Comment (ionenwks) [reply dev-qt/qtwebengine/qtwebengine-6.9999.ebuild:173]: Or wait it works now, wonder how that happened before :thinking: Well, there's still nested cmake calls that will require passing it differently though.
```
-- Found Ninja: /usr/bin/samu (found suitable version "1.9", minimum required is "1.7.2") 
```

**Comment by ionenwks** [comment]
Now with more qtwebengine that works at least on llvm-musl profiles (musl-1.2.4 is fine too) assuming I didn't break it since last test.

At this point I did all the "major" things I wanted to do (for this PR anyway) and it's mergeable'ish -- not that this is happening yet, but testing is welcome.

Rest is mostly going to be bugfixes (cpu intrinsic stuff is especially a headache) and self-review. Still need to go back over commits for potential silly mistakes (esp. with all the amending and commit count is piling up). Haven't really tried live ebuilds for status yet too.

**Comment by ionenwks** [comment]
Think ready to merge (won't be doing any notable changes anymore at least, admit I need a break), likely merge tomorrow after a last full build if nothing comes up.

Noted that before, but I do expect there will be a few confused users on non-desktop profiles that been setting opengl/vulkan per-package rather than global due to a few new `[opengl=,vulkan=,gles2-only=]`. Portage gives horrible suggestions of disabling opengl even if a  package depends on [opengl] in the match= chain. Fortunately I don't think we have that many Qt6 users yet, and I don't think it's so bad for new installs of it.
